pkg/controllers/nodeclaim/tagging/controller.go

```diff
@@ -0,0 +1,179 @@
+	timer         *time.Timer
+}
+
+func NewController(kubeClient client.Client, sess *session.Session) *Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		ec2:           ec2.New(sess),
+		defaultTagRgx: regexp.MustCompile(corev1beta1.NodePoolLabelKey + "/.*"),
+		timer:         time.NewTimer(tagPeriod),
+	}
+}
+
+func (c *Controller) Name() string {
+	return "ec2-instance-tagger"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	nodeClaim := &corev1beta1.NodeClaim{}
+	if err := c.kubeClient.Get(ctx, req.NamespacedName, nodeClaim); err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Extract resource ID from node spec
+	resourceID := path.Base(nodeClaim.Status.ProviderID)
+	if resourceID == "." || resourceID == "/" {
+		return reconcile.Result{}, fmt.Errorf("failed to extract resource ID from providerID")
+	}
+
+	// Check if instance has already been tagged
+	if isTagged, err := c.isInstanceTagged(ctx, nodeClaim, resourceID); err != nil || isTagged {
+		return reconcile.Result{}, err
+	}
+
+	// Tag instance
+	err := c.tagInstance(ctx, nodeClaim, resourceID)
+	if err != nil {
+		return reconcile.Result{Requeue: true}, client.IgnoreNotFound(err)
+	}
+
+	logging.FromContext(ctx).
+		With("node", req.NamespacedName.Name).
+		With("instanceID", resourceID).
+		Debugf("updated ec2 instance Name tag")
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(
+		controllerruntime.
+			NewControllerManagedBy(m).
+			For(&corev1beta1.NodeClaim{}).
+			WithEventFilter(onlyTaggableInstances()),
+	)
+}
+
+func onlyTaggableInstances() predicate.Predicate {
+	return predicate.NewPredicateFuncs(func(object client.Object) bool {
+		nodeClaim := object.(*corev1beta1.NodeClaim)
+
+		// Instance has already been tagged
+		if val, ok := nodeClaim.Annotations[v1beta1.AnnotationInstanceTagged]; ok && val == "true" {
+			return false
+		}
+
+		// Node name is not yet known
+		if nodeClaim.Status.NodeName == "" {
+			return false
+		}
+
+		return true
+	})
+}
+
+func (c *Controller) isInstanceTagged(ctx context.Context, nodeClaim *corev1beta1.NodeClaim, resourceID string) (bool, error) {
+	tag, err := c.getTagValue(resourceID, "Name")
+	if err != nil {
+		return false, err
+	}
+
+	// Check if the instance has the default tag (karpenter.sh/provisioner-name/*). If so it has either:
+	//   a) Already been tagged by us but it is not reflected on the API server
+	//   b) Tagged by some other process. In this case we will respect that tag.
+	if !c.defaultTagRgx.MatchString(tag) {
+		nodeClaim.Annotations[v1beta1.AnnotationInstanceTagged] = "true"
+		err := c.kubeClient.Update(ctx, nodeClaim)
+		return true, client.IgnoreNotFound(err)
+	}
+
+	return false, nil
+}
+
+func (c *Controller) getTagValue(resourceID string, tagKey string) (string, error) {
```

Comment:
Is there a way that we could reference some of the existing tag calls or abstract some of the existing tag calls using the instance provider?

Reply:
Also, if we do a `Describe` call like this, we don't get the benefit of our current batching logic on our API calls. If we use our [`Get` call on our instance provider](https://github.com/aws/karpenter/blob/main/pkg/providers/instance/instance.go#L128), we are re-using existing logic _and_ if we ever decide to increase the number of `DescribeInstance` calls we do later, we get the benefit of batching